### ISSUE TITLE
Problem compiling transmission 4.0.6

### ISSUE THREAD

**Issue by giovanni5** [transmission/transmission#6873]
### What is the issue?

I downloaded the tar.xz release and tried to compile on a raspberry pi and got this error (the same process went smoth with the previous release 4.0.5)

```
Make Error at /usr/share/cmake-3.25/Modules/ExternalProject.cmake:3115 (message):
  No download info given for 'miniupnpc' and its source directory:

   /home/pi/transmission-4.0.6/third-party/miniupnpc

  is not an existing non-empty directory.  Please specify one of:

   * SOURCE_DIR with an existing non-empty directory
   * DOWNLOAD_COMMAND
   * URL
   * GIT_REPOSITORY
   * SVN_REPOSITORY
   * HG_REPOSITORY
   * CVS_REPOSITORY and CVS_MODULE
Call Stack (most recent call first):
  /usr/share/cmake-3.25/Modules/ExternalProject.cmake:4185 (_ep_add_download_command)
  cmake/TrMacros.cmake:170 (ExternalProject_Add)
  CMakeLists.txt:505 (tr_add_external_auto_library)
```
I verified that the directory  /home/pi/transmission-4.0.6/third-party/miniupnpc  exists and it's not empty.

To compile I used this script:

```
wget -qO- https://github.com/transmission/transmission/releases/download/4.0.6/transmission-4.0.6.tar.xz | tar xJv
cd transmission-4.0.6
cmake -B build -DCMAKE_BUILD_TYPE=Release
cd build
cmake --build . -j 4
sudo cmake --install .
```

### Which application of Transmission?

transmission-daemon

### Which version of Transmission?

4.0.6

**Comment by hoimic** [comment]
In the following commit, miniupnpc was made part of miniupnp (
* refactor: add miniupnp 2.3.0 as a submodule
* refactor: remove miniupnpc submodule
):
https://github.com/transmission/transmission/commit/a72a93764e556da6b45d086162e8bc941277ab0d

It seems that the corresponding changes to CMakeLists.txt are not in the CMakeLists.txt file contained in transmission-4.0.6.tar.xz.

**Comment by giovanni5** [comment]
Yes, I modified the CMakeLists.txt file and it started compiling. The tar.xz archive should be corrected. Thanks.

**Comment by tearfur** [comment]
@ckerr Looks like the submodules are not synced to `4.0.x` when the release tar was prepared.

**Comment by tearfur** [comment]
@transmission/core-devs

**Comment by mreid-tt** [comment]
Getting this build error as well attempting to package for Synology. Any expected timeframe for a fix to be released?

**Comment by mreid-tt** [comment]
> Yes, I modified the CMakeLists.txt file and it started compiling. The tar.xz archive should be corrected. Thanks.

Would it be possible to share the specific modifications made?

**Comment by giovanni5** [comment]
> > Yes, I modified the CMakeLists.txt file and it started compiling. The tar.xz archive should be corrected. Thanks.
> 
> Would it be possible to share the specific modifications made?

It's not so easy, I solved for miniunpc, but get the same error for other libraries. In the end I just skipped compiling those libraries by installing them with APT before compiling transmission. The dependencies are listed in the instructions for compiling from source.

**Comment by tearfur** [comment]
ping @transmission/core-devs 

Well, for the time being, you can find the changes we made to the CMake files to compile with the new `miniupnp` submodule at a72a93764e556da6b45d086162e8bc941277ab0d. Unfortunately I cannot do anything about the release itself.

**Comment by Coeur** [comment]
Only @mikedld or @ckerr can do a release packaging, no one else. Assigning the issue to them.

**Comment by mreid-tt** [comment]
For those interested, @hgy59 created a source archive that includes the submodules and added it to SynoCommunity releases in https://github.com/SynoCommunity/spksrc/releases/tag/transmission. (Ignore the source code archives)

I've used this to successfully build a release for Synology. He also provided the script used to create the archive for documentation and content validation.

**Comment by dechamps** [comment]
To avoid this kind of problem in the future, would it make sense to have a GitHub Actions workflow to prepare a release?

For example, when pushing a tag that matches a release tag regex, the GitHub Actions workflow could prepare the release assets, and then use something like [`action-gh-release`](https://github.com/softprops/action-gh-release) to create a draft release (without publishing it) with the assets already pre-attached. [Example here](https://github.com/dechamps/FlexASIO/blob/e75cc72483a4f0a4cb906bda80463b0c4bd6e8d3/.github/workflows/continuous-integration.yml#L105-L124). All a human has to do then is to fill out the release description and then click publish. This reduces the potential for human error down to practically zero.

As a bonus one could also add a workflow step to generate [SLSA provenance](https://github.com/slsa-framework/slsa-github-generator/blob/main/internal/builders/generic/README.md), thereby guaranteeing the authenticity of release artifacts and making Transmission [SLSA Build Level 3](https://slsa.dev/spec/v1.0/levels#build-l3-hardened-builds) compliant.

**Comment by dechamps** [comment]
> @dechamps Could you please open a new issue for it? I think this should be more visible.

Sure thing: #7090

**Comment by gxfr** [comment]
same thing here can't install because of

`CMake Warning:
  No source or binary directory provided.  Both will be assumed to be the
  same as the current working directory, but note that this warning will
  become a fatal error in future CMake releases.


CMake Error: The source directory "/home/donnie/transmission-4.0.6/build" does not appear to contain CMakeLists.txt.
Specify --help for usage, or press the help button on the CMake GUI.`

**Comment by gxfr** [comment]
same thing here can't install because of

```CMake Warning:
  No source or binary directory provided.  Both will be assumed to be the
  same as the current working directory, but note that this warning will
  become a fatal error in future CMake releases.


CMake Error: The source directory "/home/NAME/transmission-4.0.6/build" does not appear to contain CMakeLists.txt.
Specify --help for usage, or press the help button on the CMake GUI.```

**Comment by nevack** [comment]
> same thing here can't install because of
> 
> ```cmake
>   No source or binary directory provided.  Both will be assumed to be the
>   same as the current working directory, but note that this warning will
>   become a fatal error in future CMake releases.
> 
> 
> CMake Error: The source directory "/home/NAME/transmission-4.0.6/build" does not appear to contain CMakeLists.txt.
> Specify --help for usage, or press the help button on the CMake GUI.```
> ```

This seems unrelated.
You should use `cmake ..` from inside `build/` dir.

**Comment by reinisb** [comment]
Has this been resolved? Will building 4.0.6 on Raspberry Pi OS (Debian 12) work?

**Comment by hoimic** [comment]
> Has this been resolved? Will building v4.0.6 following [these intstructions](https://github.com/transmission/transmission/blob/main/docs/Building-Transmission.md#debian-12--bookworm) on Raspberry Pi OS (Debian 12) work? I want to make sure before I uninstall v3 and go for v4.

The archive in the releases section is still from May 30th, so it has not been resolved I'm afraid. If you are okay with modifying some files before compiling and installing on your system, [take a look at this comment](https://github.com/transmission/transmission/issues/6873#issuecomment-2155998677). If you want to install a release that you don't have to modify first, you can install version 4.0.5.

**Comment by giovanni5** [comment]
As @hoimic  says, the problem is not solved, but the error occurs when building the dependencies. Luckily, this is not necessary—you can install those dependencies with apt before compiling Transmission, as reported [here](https://github.com/transmission/transmission/blob/main/docs/Building-Transmission.md).

`$ sudo apt install libb64-dev libdeflate-dev libevent-dev libminiupnpc-dev libnatpmp-dev libpsl-dev libsystemd-dev`

Should I close this issue as completed?

**Comment by reinisb** [comment]
Thanks both, I'll give this a shot! Btw if I just want to run transmission through web-interface (my raspberry py os is headless), do I need to specify anything particular when [building](https://github.com/transmission/transmission/blob/main/docs/Building-Transmission.md#building-transmission-from-git-first-time)?

**Comment by hoimic** [comment]
> Thanks both, I'll give this a shot! Btw if I just want to run transmission through web-interface (my raspberry pi os is headless), do I need to specify anything particular when [building](https://github.com/transmission/transmission/blob/main/docs/Building-Transmission.md#building-transmission-from-git-first-time)?

I use these cmake command line options to turn off all GUI components and enable building the web interface: 
`-DCMAKE_BUILD_TYPE=RelWithDebInfo -DENABLE_DAEMON=ON -DENABLE_GTK=OFF -DENABLE_QT=OFF -DINSTALL_WEB=ON -DENABLE_CLI=OFF -DENABLE_UTILS=ON`

**Comment by hoimic** [comment]
It sounds like a frustrating experience! If the compilation and installation process was completed correctly, transmission-daemon should be in PATH, so you can just start it with
`transmission-daemon --logfile ~/.local/log/transmission.log`
The full procedure for running Transmission headless [is described here](https://github.com/transmission/transmission/blob/main/docs/Headless-Usage.md).

**Comment by uiolee** [comment]
I had the same problem that confused me for a while. 

If it can't be fixed right away, why not make some changes so as not to keep people stuck. 
For example, it wouldn't be hard to add a few hints to the website, github or in release tarball.

**Comment by ckerr** [comment]
This is a long ticket but if I'm reading it correctly, this was a problem with the tarball that was made for that particular release?

If so, this should be fixed as a side-effect of the new 4.1.0 release, so I'm closing this ticket. If the problem persists in >= 4.1.0, please open a new ticket. Thank you!
